utils/stm-to-jitx.stanza

```diff
@@ -0,0 +1,276 @@
+              usart $
+                for option in options map :
+                  print("%_ " % [option])
+                  switch(option) :
+                    "UART-DTR" : UART-DTR
+                    "UART-CTS" : UART-CTS
+                    "UART-DCD" : UART-DCD
+                    "UART-RI" : UART-RI
+                    "UART-DST" : UART-DST
+                    "UART-RTS" : UART-RTS
+                    "UART-CK" : UART-CK
+                    "UART-DE" : UART-DE
+                    "UART-CS" : UART-CS
+                    "UART-RX" : UART-RX
+                    "UART-TX" : UART-TX
+            "spi" : spi $
+              for option in options map :
+                switch(option) :
+                  "SPI-SDO" : SPI-SDO
+                  "SPI-SDI" : SPI-SDI
+                  "SPI-CS" : SPI-CS
+                  "SPI-CIPO" : SPI-CIPO
+                  "SPI-COPI" : SPI-COPI
+                  "SPI-SDIO" : SPI-SDIO
+            "spi-controller" :
+              spi-controller()
+            else :
+              fatal("Bundle '%_' is not supported." % [bundle-name])
+    match(bundle) :
+      (b:Bundle) : generate-support(b, mappings(support))
+      ([b1, b2]:[Bundle, Bundle]) : generate-support(b1, b2, mappings(support))
+
+;Generate a module that accepts user options and instantiates the given component.
+public defn mcu-module (component:Instantiable) -> (Tuple<KeyValue<Symbol,?>> -> Instantiable) :
+  pcb-module module (options:Tuple<KeyValue<Symbol,?>>) :
+    val settings = Settings(DEFAULT-SETTINGS)
+    for entry in options do :
+      settings[key(entry)] = value(entry)
+      
+    public inst mcu : component
+
+    connect-reset(mcu)
+    connect-power(mcu)
+    generate-bypass(mcu, settings[`bypass-package], settings[`bypass-pin])
+    set-boot(mcu, settings[`boot-from])
+    connect-debug(mcu, settings[`debug-interface], settings[`debug-connector])
+    setup-clocks(mcu, settings[`HSE-freq], settings[`HSE-ppm], settings[`HSE-source], settings[`LSE-freq], settings[`LSE-ppm], settings[`LSE-source])
+
+    schematic-group(self) = stm
+  module
+
+;=========================================
+;=============== Utilities ===============
+;=========================================
+
+;Convert a String that may have brackets to a Ref.
+defn ref-string-to-ref (text:String) -> Ref :
+  ;regex-match implementation incomplete so we can't do exact matches with "^mypattern$"
+  ;val match-result = regex-match("^(\\w+).([\\w\.])+$", text)
+  if text[length(text) - 1] == ']' :
+    val idx = last-index-of-char(text, '[') as Int
+    IndexRef(ref-string-to-ref(text[0 to idx]), to-int!(text[(idx + 1) to (length(text) - 1)]))
+  else :
+    val idx = last-index-of-char(text, '.')
+    match(idx: Int) :
+      FieldRef(ref-string-to-ref(text[0 to idx]), Ref(text[(idx + 1) to length(text)]))
+    else :
+      Ref(text)
+
+;Convert a String without brackets to a Ref.
+defn raw-string-to-ref (text:String) -> Ref :
```

Comment:
Okay the implementation of `last-index-of-char` surely does a loop but it is hidden away for better readability

Reply:
These two functions have different behaviors (hinted at by the comments, although they could be more descriptive).
`ref-string-to-ref` will turn `"Q[14]"` into `IndexRef(Ref("Q"), 14)`
`raw-string-to-ref` assumes the given string does not have any `[` `]` and will not behave as you'd expect on `"Q[14]"`. It would return `Ref("Q[14]")`.

`raw-string-to-ref` will turn `"Q14"` into `IndexRef(Ref("Q"), 14)`
`ref-string-to-ref` expects brackets to already exist for IndexRef. It would return `Ref("Q[14]")`.

Reply:
I don't think we want any ref of the form `Ref("Q[14]")`, do you have examples? We can rename them in the mongodb so that this doesn't happen